### PR TITLE
Try std::unique_ptr for ClingATSImporter

### PR DESCRIPTION
**Pull request by CristinaCristescu**

### CODE REVIEW

File: interpreter/cling/lib/Interpreter/ExternalInterpreterSource.cpp [hunk 1]
```diff
@@ -65,8 +66,9 @@ namespace {
 namespace cling {
 
   ExternalInterpreterSource::ExternalInterpreterSource(
-          const cling::Interpreter *parent, cling::Interpreter *child) :
-          m_ParentInterpreter(parent), m_ChildInterpreter(child) {
+        const cling::Interpreter *parent, cling::Interpreter *child) :
+        m_ParentInterpreter(parent), m_ChildInterpreter(child),
+        m_Importer(nullptr) {
```
Comment (karies) [inline interpreter/cling/lib/Interpreter/ExternalInterpreterSource.cpp]: Using the default initialization of `m_Importer` is enough.

File: interpreter/cling/lib/Interpreter/ExternalInterpreterSource.cpp [hunk 2]
```diff
@@ -76,25 +78,21 @@ namespace cling {
...
-    FileManager &parentFM = parent->getCI()->getFileManager();
+    FileManager &childFM = m_ChildInterpreter->getCI()->getFileManager();
+    FileManager &parentFM = m_ParentInterpreter->getCI()->getFileManager();
 
-    return ClingASTImporter(toContext, childFM, fromContext, parentFM,
-                            /*MinimalImport : ON*/ true, source);
+    ASTContext &fromASTContext = m_ParentInterpreter->getCI()->getASTContext();
+    ASTContext &toASTContext = m_ChildInterpreter->getCI()->getASTContext();
+    ClingASTImporter* importer = new ClingASTImporter(toASTContext, childFM, 
+                                          fromASTContext, parentFM,
```
Comment (karies) [inline interpreter/cling/lib/Interpreter/ExternalInterpreterSource.cpp]: indentation?

File: interpreter/cling/lib/Interpreter/ExternalInterpreterSource.cpp [hunk 3]
```diff
@@ -107,7 +105,8 @@ namespace cling {
         && declToImport->isTemplateDecl())
       return;
 
-    if (Decl *importedDecl = importer.Import(declToImport)) {
+    //ClingASTImporter* clingImporter = llvm::dyn_cast<ClingASTImporter>(m_Importer);
+    if (Decl *importedDecl = m_Importer.get()->Import(declToImport)) {
```
Comment (karies) [inline interpreter/cling/lib/Interpreter/ExternalInterpreterSource.cpp]: `m_Importer->Import` should work just fine?

File: interpreter/cling/lib/Interpreter/ExternalInterpreterSource.cpp [hunk 4]
```diff
@@ -121,13 +120,13 @@ namespace cling {
...
                                   DeclContext *declContextToImport,
-                                  ASTImporter &importer,
                                   DeclarationName &childDeclName,
                                   DeclarationName &parentDeclName,
                                   const DeclContext *childCurrentDeclContext) {
 
+    //ClingASTImporter clingImporter = dyn_cast<ClingASTImporter>(m_Importer);
     if (DeclContext *importedDeclContext =
-                                importer.ImportContext(declContextToImport)) {
+                                m_Importer.get()->ImportContext(declContextToImport)) {
```
Comment (karies) [inline interpreter/cling/lib/Interpreter/ExternalInterpreterSource.cpp]: Ditto

File: interpreter/cling/lib/Interpreter/ExternalInterpreterSource.cpp [hunk 5]
```diff
@@ -148,30 +147,23 @@ namespace cling {
...
       // a DeclContext (for example a Namespace, function etc.).
       if (DeclContext *declContextToImport = llvm::dyn_cast<DeclContext>(*I)) {
 
-        ImportDeclContext(declContextToImport, importer, childDeclName,
+        ImportDeclContext(declContextToImport, childDeclName,
                           parentDeclName, childCurrentDeclContext);
 
       }
-      ImportDecl(*I, importer, childDeclName, parentDeclName,
+      ImportDecl(*I, childDeclName, parentDeclName,
```
Comment (karies) [inline interpreter/cling/lib/Interpreter/ExternalInterpreterSource.cpp]: One line?

File: interpreter/cling/lib/Interpreter/ExternalInterpreterSource.cpp [hunk 6]
```diff
@@ -279,7 +255,7 @@ namespace cling {
         if (auto II = childDeclName.getAsIdentifierInfo()) {
           StringRef name = II->getName();
           if (!name.empty() && name.startswith(filter))
-            ImportDecl(parentDecl, importer, childDeclName, childDeclName,
+            ImportDecl(parentDecl, childDeclName, childDeclName,
```
Comment (karies) [inline interpreter/cling/lib/Interpreter/ExternalInterpreterSource.cpp]: One line? (I cannot see whether that's < 80 - this is an actual question :-)

**Comment by karies** [comment]
Nice! Thanks! Could you address the few comments I had?

**Comment by CristinaCristescu** [comment]
I think I addressed them all and ctest is happy!

File: interpreter/cling/lib/Interpreter/ExternalInterpreterSource.cpp (hunk 6 above)
Comment (CristinaCristescu) [inline interpreter/cling/lib/Interpreter/ExternalInterpreterSource.cpp]: It's 83..:(

File: interpreter/cling/lib/Interpreter/ExternalInterpreterSource.cpp [hunk 7]
```diff
@@ -29,7 +29,8 @@ namespace {
                      bool MinimalImport, cling::ExternalInterpreterSource& source):
       ASTImporter(ToContext, ToFileManager, FromContext, FromFileManager,
                   MinimalImport), m_Source(source) {}
-    virtual ~ClingASTImporter() = default;
+
+      virtual ~ClingASTImporter() = default;
```
Comment (karies) [inline interpreter/cling/lib/Interpreter/ExternalInterpreterSource.cpp]: Indentation?

Comment (CristinaCristescu) [inline interpreter/cling/lib/Interpreter/ExternalInterpreterSource.cpp]: Should be ok now

**Comment by karies** [comment]
Thanks, merged!

**Pull request closed without merging**